File: Python/ByRound/1860/1860_F_Evaluate_RBS.py
# Problem: CF 1860 F - Evaluate RBS
# https://codeforces.com/contest/1860/problem/F

"""
Problem: Evaluate RBS
Algorithm: 
    - This problem involves determining if we can sort a set of bracket tuples such that they form a valid bracket sequence.
    - We use a geometric approach where each tuple (a, b, c) is treated as a point in a 2D plane.
    - For each pair of tuples, we compute the intersection point of their "lines" to determine orderings.
    - The approach checks valid orderings using a sweep-line like technique with coordinate compression.

Time Complexity: O(n^3 * log(n)) where n is the number of pairs of brackets (2n total tuples)
Space Complexity: O(n^2) for storing intermediate data structures
"""

import bisect
import io
import os
import sys
from collections import defaultdict

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v, w, x):
    # Helper function to create a unique hash value for tuple (u,v,w,x)
    return u << 40 ^ v << 20 ^ w << 12 ^ x


def g(u, v):
    # Helper function to combine two indices into one value
    return u << 12 ^ v


t = int(input())
ans = []
for _ in range(t):
    n = int(input())
    a, b, c, x = [0] * (2 * n), [0] * (2 * n), [0] * (2 * n), [0] * (2 * n)
    for i in range(2 * n):
        a0, b0, c0 = input().rstrip().decode().split()
        a0, b0, c0 = int(a0), int(b0), ord(c0) & 1  # Convert char to 0 or 1
        a[i], b[i], c[i] = a0, b0, c0
        x[i] = f(a0, b0, c0, i)
    
    # Sort using custom hash and get indices
    u = [i & 0xFFF for i in sorted(x)]
    v = [0] * (2 * n)
    for i in range(2 * n):
        v[u[i]] = i
    
    # Find all intersection points of lines formed by tuples
    t1 = []
    for i in range(2 * n):
        ai, bi = a[i], b[i]
        for j in range(i + 1, 2 * n):
            aj, bj = a[j], b[j]
            if ai == aj or bi == bj:
                continue
            t0 = (ai - aj) / (bj - bi)
            if t0 > 0:
                t1.append(t0)
    
    t1.sort()  # Sort intersection points
    lt = len(t1)
    d = [0] * lt
    w = [i for i in range(lt)]
    
    # Map these intersection points to index pairs
    for i in range(2 * n):
        ai, bi = a[i], b[i]
        for j in range(i + 1, 2 * n):
            aj, bj = a[j], b[j]
            if ai == aj or bi == bj:
                continue
            t0 = (ai - aj) / (bj - bi)
            if t0 > 0:
                k = bisect.bisect_left(t1, t0)
                d[w[k]] = g(i, j)
                w[k] += 1
    
    # Precompute prefix sums for bracket counts
    s = [0] * (2 * n + 1)
    c = [c[u[i]] for i in range(2 * n)]
    for i in range(2 * n):
        s[i + 1] = s[i] + 1 if not c[i] else s[i] - 1
    
    cnt = [0] * (3 * n)
    for i in s:
        cnt[i] += 1
    
    y = [0] * (2 * n)
    ok = 0
    l = 0
    # Process each group of intersection points
    while l ^ lt and cnt[-1]:
        x = defaultdict(lambda: 0)
        r = w[l]
        for i in range(l, r):
            j, k = d[i] >> 12, d[i] & 0xFFF
            y[v[j]], y[v[k]] = j, k
            if v[j] < v[k]:
                x[v[j]] += 1
                x[v[k]] += 0
            else:
                x[v[j]] += 0
                x[v[k]] += 1
        l = r
        c1 = 0
        for i in x.keys():
            if i + 1 in x and x[i] >= x[i + 1]:
                if s[i + 1] == -1:
                    c1 += 1
            elif s[i + 1] < 0:
                c1 = -1
                break
        if c1 == cnt[-1]:
            ok = 1
            break
        st = []
        for i in sorted(x.keys()):
            st.append(y[i])
            if i + 1 in x and x[i] == x[i + 1]:
                continue
            while st:
                j = v[st.pop()]
                k = j + 1
                for _ in range(x[i]):
                    cnt[s[k]] -= 1
                    if c[j] < c[k]:
                        s[k] -= 2
                    elif c[j] > c[k]:
                        s[k] += 2
                    cnt[s[k]] += 1
                    u[j], u[k] = u[k], u[j]
                    c[j], c[k] = c[k], c[j]
                    v[u[j]], v[u[k]] = v[u[k]], v[u[j]]
                    j, k = j + 1, k + 1
    if not cnt[-1]:
        ok = 1
    ans0 = "YES" if ok else "NO"
    ans.append(ans0)
sys.stdout.write("\n".join(ans))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1860/1860_F_Evaluate_RBS.cpp
// Problem: CF 1860 F - Evaluate RBS
// https://codeforces.com/contest/1860/problem/F

/*
Algorithm: This solution attempts to determine if there exist positive x and y such that sorting the brackets
by a·x + b·y results in a valid bracket sequence. The key insight is that for a valid bracket sequence,
the prefix sum of bracket values must not go negative at any point, and must end at zero.

Approach:
1. Process input data to build arrays of bracket data.
2. Preprocess the bracket sequence to compute prefix sums and track minimum prefix sum.
3. Generate candidate swaps based on differences in (a,b) values and bracket types.
4. Sort these candidates by a custom comparator that uses cross products to avoid floating point arithmetic.
5. Iterate through sorted candidates and attempt to make swaps that reduce or maintain the prefix sum.
6. Check if we can make the total prefix sum become zero, making the sequence valid.

Time Complexity: O(n^4) in worst case due to nested loops in preprocessing and swap attempts.
Space Complexity: O(n^2) for storing candidate swaps and intermediate arrays.
*/

#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
const int c = 3005;
long long n, l[c], r[c], ert[c]; // l: 'a' values, r: 'b' values, ert: bracket values (-1 for '(' , 1 for ')')
int t[c], inv[c], pref[c], sum, jo; // t: sorting permutation, inv: inverse permutation, pref: prefix sums
vector<pair<pair<int, int>, int>> sz; // sz: list of candidates for swaps
bool cmp(pair<pair<int, int>, int> a, pair<pair<int, int>, int> b) {
  // Custom comparator to avoid floating point arithmetic using cross product
  int A = a.first.first, B = a.first.second, C = b.first.first,
      D = b.first.second;
  long long x = (l[A] - l[B]) * (r[D] - r[C]), // Cross product to compare slopes
            y = (l[C] - l[D]) * (r[B] - r[A]);
  return (x < y || (x == y && a.second < b.second)); // Return true if a should come before b
}
void solve() {
  cin >> n;
  n *= 2; // Total number of brackets (n open, n close)
  for (int i = 1; i <= n; i++) {
    cin >> l[i] >> r[i];
    char c;
    cin >> c;
    ert[i] = (c == '(' ? -1 : 1); // Assign bracket value
  }
  // Sort tuples by a,b values (in this case, simple sorting for simplicity)
  for (int i = 1; i <= n; i++) {
    for (int j = i + 1; j <= n; j++) {
      if (l[j] < l[i] || (l[j] == l[i] &&
                          (r[j] < r[i] || (r[j] == r[i] && ert[j] < ert[i])))) {
        swap(l[i], l[j]);
        swap(r[i], r[j]);
        swap(ert[i], ert[j]);
      }
    }
  }
  // Set up initial prefix sums and inverse permutation arrays
  for (int i = 1; i <= n; i++) {
    t[i] = i, inv[i] = i;
    pref[i] = pref[i - 1] + ert[i];
    sum += max(0, pref[i]); // Sum of positive prefix values
  }
  jo = (sum == 0); // Check if initial total is valid
  // Generate candidate swaps based on bracket values and positions
  for (int i = 1; i <= n; i++) {
    for (int j = i + 1; j <= n; j++) {
      if (r[j] < r[i]) { // If right value of bracket j is less than bracket i
        sz.push_back({{i, j}, ert[j] - ert[i]}); // Add candidate swap with bracket difference
      }
    }
  }
  sort(sz.begin(), sz.end(), cmp); // Sort candidates using custom comparator
  for (auto x : sz) {
    int A = x.first.first, B = x.first.second, val = x.second;
    if (inv[A] < inv[B]) { // If bracket A is before B in current order
      swap(t[inv[A]], t[inv[B]]); // Swap their positions in the sort order
      for (int i = inv[A]; i < inv[B]; i++) {
        sum -= max(0, pref[i]);
        pref[i] += val;
        sum += max(0, pref[i]); // Update prefix sum changes
      }
      swap(inv[A], inv[B]); // Update inverse permutation
    }
    if (sum == 0) {
      jo = 1; // If valid, set solution flag
    }
  }
  cout << (jo ? "YES" : "NO") << "\n";
  // Reset arrays for next test case
  for (int i = 0; i < c; i++) {
    l[i] = 0, r[i] = 0, ert[i] = 0, t[i] = 0, inv[i] = 0, pref[i] = 0;
  }
  jo = 0, sum = 0;
  sz.clear();
}
int main() {
  ios_base::sync_with_stdio(false);
  int w;
  cin >> w;
  while (w--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/